packages/ai/ai/src/Embeddings.ts

```diff
@@ -0,0 +1,152 @@
+
+const makeBatchedResolver = (
+  embedMany: (input: ReadonlyArray<string>) => Effect.Effect<Array<Embeddings.Result>, AiError>
+) =>
+  RequestResolver.makeBatched(
+    (requests: ReadonlyArray<EmbeddingRequest>) =>
+      embedMany(requests.map((request) => request.input)).pipe(
+        Effect.flatMap(
+          Effect.forEach(
+            ({ embeddings, index }) => Request.succeed(requests[index], embeddings),
+            { discard: true }
+          )
+        ),
+        Effect.catchAll((error) =>
+          Effect.forEach(
+            requests,
+            (request) => Request.fail(request, error),
+            { discard: true }
+          )
+        )
+      )
+  )
+
+/**
+ * @since 1.0.0
+ * @category constructors
+ */
+export const make = (options: {
+  readonly embedMany: (input: ReadonlyArray<string>) => Effect.Effect<Array<Embeddings.Result>, AiError>
+  readonly maxBatchSize?: number
+  readonly cache?: {
+    readonly capacity: number
+    readonly timeToLive: Duration.DurationInput
+  }
+}) =>
+  Effect.gen(function*() {
+    const cache = yield* Option.fromNullable(options.cache).pipe(
+      Effect.flatMap((config) => Request.makeCache(config)),
+      Effect.optionFromOptional
+    )
+
+    const resolver = makeBatchedResolver(options.embedMany).pipe(
+      options.maxBatchSize ? RequestResolver.batchN(options.maxBatchSize) : identity
+    )
+
+    function embed(input: string) {
+      const request = Effect.request(new EmbeddingRequest({ input }), resolver)
+      return Option.match(cache, {
+        onNone: () => request,
+        onSome: (cache) =>
+          request.pipe(
+            Effect.withRequestCaching(true),
+            Effect.withRequestCache(cache)
+          )
+      }).pipe(Effect.withSpan("Embeddings.embed", {
+        captureStackTrace: false,
+        attributes: { input }
+      }))
+    }
+
+    return Embeddings.of({
+      embed
+    })
+  }).pipe(Effect.withSpan("Embeddings.make"))
+
+/**
+ * Creates an `Embeddings` service which will aggregate all `embed` requests
+ * received during the specified `window` (up to a maximum of `maxBatchSize`
+ * requests, if specified) and execute them as a single batch.
+ *
+ * @since 1.0.0
+ * @category constructors
+ */
+export const makeDataLoader = (options: {
+  readonly embedMany: (input: ReadonlyArray<string>) => Effect.Effect<Array<Embeddings.Result>, AiError>
+  readonly window: Duration.DurationInput
+  readonly maxBatchSize?: number
+}) =>
+  Effect.gen(function*() {
+    const scope = yield* Effect.scope
+
+    const resolver = makeBatchedResolver(options.embedMany)
+    const resolverDelayed = yield* dataLoader(resolver, {
+      window: options.window,
+      maxBatchSize: options.maxBatchSize
+    })
+
+    function embed(input: string) {
+      return Effect.request(new EmbeddingRequest({ input }), resolverDelayed).pipe(
+        Effect.withSpan("Embeddings.embed", {
+          captureStackTrace: false,
+          attributes: { input }
```

Comment:
Probably not a good idea to attach the input to the span. It could be quite large.